1017 python aws (#32)
signalfx/lambda-python: README.rst

SignalFx Python Lambda Wrapper
===============================

Overview
---------

You can use this document to add a SignalFx wrapper to your AWS Lambda for Python. 

The SignalFx Python Lambda Wrapper wraps around an AWS Lambda Python function handler, which will allow metrics and traces to be sent to SignalFx.

At a high-level, to add a SignalFx python lambda wrapper, you can either package it yourself or you can use a Lambda layer containing the wrapper, and then attach the layer to a Lambda function.

Step 1: Add the Lambda wrapper in AWS
-----------------------------------------

To add the SignalFx wrapper, you have the following options:
   * Option 1: Use the wrapper as a regular dependency, and then create a Lambda function based on your artifact containing both code and dependencies.
   * Option 2: In AWS, create a Lambda function, then attach a SignalFx-hosted layer with a wrapper.

>    * Option 3: In AWS, create a Lambda function, then create and attach a layer based on a SignalFx SAM (Serverless Application Model) template.

Reviewer: I think also provide a link to Lambda Layers when presenting the option for people who are new to them (they are a relatively new concept).

>    * Option 3: In AWS, create a Lambda function, then create and attach a layer based on a SignalFx SAM (Serverless Application Model) template.

Author: I added the link in the "overview" section. Most importantly, I rearranged the options:

- Previous Option 1, is now Option 3. 
- Previous Option 2, is now Option 1
- Previous Option 3, is now Option 2

Review the appropriate option below.

> Option 1: Install the wrapper

Author: This has been updated.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Run the following installation script in your command line:

.. code::

    pip install signalfx_lambda

> Option 2: Create a Lambda function, then attach the SignalFx-hosted layer

Reviewer: ```suggestion
Option 2: Create a Lambda function, then attach the SignalFx-hosted Lambda layer
```

Author: This has been updated.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> In this option, you will use a layer created and hosted by SignalFx.

Reviewer: ```suggestion
In this option, you will use a Lambda layer created and hosted by SignalFx.
```

Author: This has been updated.

10. Mark **Provide a layer version**.
11. Enter an ARN number. 
  * To locate the ARN number, see [Lambda Layer Versions](https://github.com/signalfx/lambda-layer-versions/blob/master/python/PYTHON.md).

Option 3: Create a Lambda function, then create and attach a layer based on a SignalFx template
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In this option, you will chose a SignalFx template, and then deploy a copy of the layer.

1. Open your AWS console. 
2. In the landing page, under **Compute**, click **Lambda**.
3. Click **Create function** to create a layer with SignalFx's capabilities.
4. Click **Browse serverless app repository**.
5. Click **Public applications**.
6. In the search field, enter and select **signalfx-lambda-python-wrapper**.
7. Review the template, permissions, licenses, and then click **Deploy**.
    * A copy of the layer will now be deployed into your account.
8. Return to the previous screen to add a layer to the function, select from list of runtime compatible layers, and then select the name of the copy.  

Step 2: Locate and set the ingest endpoint
-------------------------------------------

> By default, this function wrapper will send data to the us0 realm. If you are not in us0 realm and you want to use the ingest endpoint directly, then you muset set your realm. 

Reviewer 2: *must set

Author: Good catch. This has been updated.

To set your realm, use a subdomain, such as ingest.us1.signalfx.com or ingest.eu0.signalfx.com.

To locate your realm:

1. Open SignalFx and in the top, right corner, click your profile icon.
2. Click **My Profile**.
3. Next to **Organizations**, review the listed realm.

Step 3: Set environment variables
----------------------------------

Please note for environment variables, SignalFx defaults to the us0 realm. As a result, if you are not in the us0 realm, you may need to set your environment variables. 

1. Set SIGNALFX_ACCESS_TOKEN with your correct access token. Review the following example. 

.. code:: bash

    SIGNALFX_ACCESS_TOKEN=access token

> 2. If you use POPS, Smart Gatway, or want to ingest directly from a realm other than us0, then you must set at least one endpoint variable. (SignalFx defaults to the us0 realm. As a result, if you are in a different realm, you must explictly set the realm.) You can update one of the varibles below. Review the following examples.  

Reviewer: ```suggestion
2. If you use POPS, Smart Gateway, or want to ingest directly from a realm other than us0, then you must set at least one endpoint variable. (SignalFx defaults to the us0 realm. As a result, if you are in a different realm, you must explicitly set the realm.) You can update one of the variables below. Review the following examples.  
```

Reviewer: hyperlinks for POPS and Smart Gateway would be nice - 

https://docs.signalfx.com/en/latest/integrations/integrations-reference/integrations.signalfx.point.of.presence.service.(pops).html

https://docs.signalfx.com/en/latest/apm/apm-deployment/smart-gateway.html

Author: Added the links, as well as corrected the misspelled 'gateway'

    pip install signalfx_lambda

Step 2: Locate and set the ingest endpoint
-------------------------------------------
By default, this function wrapper will send data to the us0 realm. As a result, if you are not in us0 realm and you want to use the ingest endpoint directly, then you must explicitly set your realm. To set your realm, use a subdomain, such as ingest.us1.signalfx.com or ingest.eu0.signalfx.com.

To locate your realm:

1. Open SignalFx and in the top, right corner, click your profile icon.
2. Click **My Profile**.
3. Next to **Organizations**, review the listed realm.

Step 3: Set environment variables
----------------------------------

1. Set SIGNALFX_ACCESS_TOKEN with your correct access token. Review the following example. 

.. code:: bash

    SIGNALFX_ACCESS_TOKEN=access token

2. If you use POPS, Smart Gateway, or want to ingest directly from a realm other than us0, then you must set at least one endpoint variable. (For environment variables, SignalFx defaults to the us0 realm. As a result, if you are not in the us0 realm, you may need to set your environment variables.) You can update one of the variables below. Review the following examples.  

> 

Reviewer 3: Hi George, sorry, I thought I wrote it somewhere;) I would put lines 52-74 back here, just after line 97. The rest should stay as it is.

7. In **Runtime**, select the desired language.
8. Click **Create function**. 
9. Click on **Layers**, then add a layer.
10. Mark **Provide a layer version**.
11. Enter an ARN number. 
  * To locate the ARN number, see [Lambda Layer Versions](https://github.com/signalfx/lambda-layer-versions/blob/master/python/PYTHON.md).

~~~~~

**Option 3: Create a Lambda function, then create and attach a layer based on a SignalFx template**

In this option, you will chose a SignalFx template, and then deploy a copy of the layer.

1. Open your AWS console. 
2. In the landing page, under **Compute**, click **Lambda**.
3. Click **Create function** to create a layer with SignalFx's capabilities.
4. Click **Browse serverless app repository**.
5. Click **Public applications**.
6. In the search field, enter and select **signalfx-lambda-python-wrapper**.
7. Review the template, permissions, licenses, and then click **Deploy**.
    * A copy of the layer will now be deployed into your account.
8. Return to the previous screen to add a layer to the function, select from list of runtime compatible layers, and then select the name of the copy.  

~~~~~

> Step 2: Update the SignalFx realm

Reviewer 3: Apologies, I unintentionally I misled you - I've been double checking various setup options and my proposed changes were not that good. 
So: the realm only matters if a user wishes to send metric directly to ingest endpoint (as opposed to using their gateway or pops). So maybe it would be better to go back on that and say sth in the lines of:

Setp 2: Locate correct ingest endpoint. 
By default, this function wrapper will send data to the ``us0`` realm. If you are not in this realm, and wish to use ingest endpoint directly, please make sure to point to the correct realm by using a subdomain, such as us1.signalfx.com ( {REALM}.signalfx.com ) - for example, ingest.us1.signalfx.com or ingest.eu0.signalfx.com.

Author: Thank you for the clarification! This paragraph has been updated.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

By default, this function wrapper will send data to the ``us0`` realm. If you are not in this realm, and you are not using the gateway, then you will need to update the realm for the SignalFx subdomain, such as us1.signalfx.com ( {REALM}.signalfx.com  ). 

To locate your realm:

1. Open SignalFx and in the top, right corner, click your profile icon.
2. Click **My Profile**.
3. Next to **Organizations**, review the listed realm.

> Additionally, you must adjust the variables you set to the correct realm ingest endpoint (``https://ingest.{REALM}.signalfx.com``). These variables include ``SIGNALFX_ENDPOINT_URL``, ``SIGNALFX_TRACING_URL``, ``SIGNALFX_METRICS_URL``, and ``SIGNALFX_INGEST_ENDPOINT``. 

Reviewer 3: I would remove line 82. With the all possible setups it is probably more confusing than helping.  I think it will be clear that if they want to use ingest directly, they need to account for the realm.

Author: This has been removed.

~~~~~

Step 3: Set environment variables
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Based on your realm, you may need to set specific environment variables. 
  * Regardless of your realm, every user must set SIGNALFX_ACCESS_TOKEN. 

>   * If you are not in the us0 realm, then you must set SIGNALFX_ENDPOINT_URL. 

Reviewer 3: What would you say about: 

If you are using pops, gateway or wish to use ingest directly from realm other than us0, you must set at least one of endpoint variables appropriately (SIGNALFX_ENDPOINT_URL or SIGNALFX_METRICS_URL or SIGNALFX_INGEST_ENDPOINT (deprecated)). See below. 

?

Author: I'm a little confused with SIGNALFX_INGEST_ENDPOINT. Is it already deprecated? If so, should we even mention it to the user? 

Please review the changes to this section. I wasn't sure if your comments were meant to replace the rest of the content in that section. 

-----
(Removed content)
You can use ``SIGNALFX_ENDPOINT_URL`` to configure a common endpoint for metrics and
traces. In other words, you can set ``SIGNALFX_ENDPOINT_URL`` to replace ``SIGNALFX_TRACING_URL`` and ``SIGNALFX_METRICS_URL``. However, if you explicitly set either ``SIGNALFX_TRACING_URL`` or ``SIGNALFX_METRICS_URL``, then ``SIGNALFX_ENDPOINT_URL`` will be overwritten.  

Review the following examples. 

If only ``SIGNALFX_ENDPOINT_URL`` is set, then both metrics and traces will be sent to the gateway address:

.. code:: bash

    SIGNALFX_ENDPOINT_URL=http://<my_gateway>:8080

If both ``SIGNALFX_ENDPOINT_URL`` and ``SIGNALFX_METRICS_URL`` are set, then traces will be sent to the gateway and metrics will go through POPS.

.. code:: bash
    
    SIGNALFX_ENDPOINT_URL=http://<my_gateway>:8080
    SIGNALFX_METRICS_URL=https://pops.signalfx.com

Reviewer 3: I only meant to replace one sentence "Based on your realm, you may need... " and the two options below :) 
SIGNALFX_INGEST_ENDPOINT can still be used, but a user will get the deprecation warning if they set it. I only wanted to leave it here in case anyone who has existing deployments/config comes back to the doc for some answers.

Author: Based on what I thought was your feedback, versus your real feedback, versus what I edit, please let me know if my new "step 3" makes sense. If not, I can easily revert to the original step 3.

    def handler(event, context):
        # your code

Step 5: Send custom metrics from the Lambda function
-------------------------------------------------------

1. Run the following command in your command line: 

.. code:: python

    import signalfx_lambda

    # sending application_performance metric with value 100 and dimension abc:def
    signalfx_lambda.send_gauge('application_performance', 100, {'abc':'def'})

    # sending counter metric with no dimension
    signalfx_lambda.send_counter('database_calls', 1)

Step 6: Add tracing to the Lambda function
-------------------------------------------

1. To trace critical parts of your handler
function, run the following command in your command line: 

.. code:: python

    import opentracing

    tracer = opentracing.tracer

    def some_function():
        with tracer.start_active_span("span_name", tags=tags) as scope:

            # do some work

            span = scope.span
            span.set_tag("example_tag", "example_value")

To review more examples and usage details, see 

> ` Jaeger Python Tracer documentation <https://github.com/signalfx/jaeger-client-python>`_.

Author: I think using parentheses and brackets should work here. 

[Jaeger Python Tracer](https://github.com/signalfx/jaeger-client-python>)

    def handler(event, context):
        # your code

Step 5: Send custom metrics from a Lambda function
-------------------------------------------------------

> 1. To send custom metrics from a Lambda function, run the following commands in your command line: 

Reviewer 3: Please change "run the following commands in your command line:" to "include the following code in your function:" or equivalent. The same remark for Step 6.1. Except for that, I am OK with merging this changes.

Author: This has been updated.

~~~~~

Step 6: Add tracing to the Lambda function
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1. To trace critical parts of your handler
function, run the following command in your command line: 

.. code:: python

    import opentracing

    tracer = opentracing.tracer

    def some_function():
        with tracer.start_active_span("span_name", tags=tags) as scope:

            # do some work

            span = scope.span
            span.set_tag("example_tag", "example_value")

To review more examples and usage details, see 
` Jaeger Python Tracer documentation <https://github.com/signalfx/jaeger-client-python>`_.

~~~~~

Step 7: Test configurations locally 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1. Run the following command in your command line: 

.. code::

    pip install python-lambda-local

.. code::

    python-lambda-local tests/test.py tests/event.json -a 'arn:aws:lambda:us-east-1:accountId:function:functionNamePython:$LATEST'

~~~~~

Step 8: Install a Python package (build a wheel)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1. Run the following command in your command line: 

.. code::

    python setup.py bdist_wheel --universal

> ~~~~~

Author: Steps 7 and 8 have been removed from the "installation" section and added to the bottom of the document.

|                                  | ‘python-lambda-wrapper’          |
+----------------------------------+----------------------------------+

> Steps 7 & 8 don't have (and shouldn't) be done by most users. This is rather info for contributors who want to test changes they do in the wrapper and/or publish a new version. As such, I would vote to have them in last separate section with a title "For contributors" or "For maintainers" or just "Additional info". The metrics and dimensions and tags which are now last are probably of interest to most users, but this two steps are not.

Reviewer 3: Looks like copy-paste error, this shouldn't be here, this was a comment.

Author: Yes, you are right! ;) I was copying / pasting your comments into the document to help me understand what action to perform. This has been updated.

> Test configurations locally 

Reviewer 3: I would just say "Test locally" without "configurations". As a programmer, I wouldn't know what "configurations" are in this context. I think we can also assume that someone who tests this locally know what they do and don't need more info.

Author: This has been updated.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> For users who are interested in testing changes to a wrapper or in publishing a new version, run the following commmand in your command line: 

Reviewer 3: I would remove "or in publishing a new version" so that this sentence just becomes:
"For users who are interested in testing changes to a wrapper, run the following command in your command line:"

Author: This has been updated.

|                                  | ‘python-lambda-wrapper’          |
+----------------------------------+----------------------------------+

Test locally 
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> For users who are interested in testing changes to a wrapper, run the following commmand in your command line: 

Reviewer: ```suggestion
For users who are interested in testing changes to a wrapper, run the following command in your command line: 
```

Author: This has been updated.

    python-lambda-local tests/test.py tests/event.json -a 'arn:aws:lambda:us-east-1:accountId:function:functionNamePython:$LATEST'

Publish a new version
^^^^^^^^^^^^^^^^^^^^^^^
If you would like to publish a new version, run the following command in your command line to install a new Python package (build a wheel): 

.. code::

    python setup.py bdist_wheel --universal
    
License
^^^^^^^^

> Apache Software License v2. Copyright © 2014-2019 SignalFx    

Reviewer 3: We made it to 2020 ;)

Author: This has been updated.

Reviewer 4: We (Joe and I) had to update our copyright statements to say Splunk. To be exact:

(c) Copyright 2020 Splunk, Inc. 

So if this type of copyright statement is in the readme files, they all need to be updated. (Fine with me if it says 2014-2020.) Need a jira ticket and should do it ASAP, obviously.

Reviewer 3: Thanks for clarification. I updated the copyrights to say Splunk.

|                                  | ‘python-lambda-wrapper’          |
+----------------------------------+----------------------------------+

~~~~~

> 

Author: This has been added back to the bottom of the document.
